In [1]:
# import my requried Dependencies
import pandas as pd
import numpy as np

In [2]:
#add files

pymoli_data = pd.read_csv("ref\\purchase_data.csv")

In [3]:
#to find the total players count the unique SN#

total_players = len(pymoli_data['SN'].unique())

#tp_df = pd.Dataframe({"total players": [total_players]})
#top_df
total_players
print("The Total Players: " + str(total_players))

#576 unique total players. duplicates must be present in the data.
#have to remove duplicates before re-running total gender count

The Total Players: 576


In [4]:
#Number of Unique Items
#Average Purchase Price
#Total Number of Purchases
#Total Revenue

In [40]:
items = len(pymoli_data['Item ID'].unique())
Number_Purchases = pymoli_data['Purchase ID'].count()
average_price = (pymoli_data['Price'].sum()) / Number_Purchases
Total_revenue = pymoli_data['Price'].sum()

# create a summary table that has columns listed below. this is created via dictionary of lists. 
summary_table_df = pd.DataFrame({'The Amount of Anique Items Purchased': [items], 
                                'Average Price': [average_price], 'Total Number of Purchases': [Number_Purchases],
                                 'Total Revenue for Pymoli': [Total_revenue]})
summary_table_df

#add format to include $ sign if column includes price
summary_table_df['Average Price'] = summary_table_df['Average Price'].map("${0:,.2f}".format)
summary_table_df['Total Revenue for Pymoli'] = summary_table_df['Total Revenue for Pymoli'].map("${0:,.2f}".format)

summary_table_df


The Amount of Anique Items Purchased Average Price  \
0                                   183         $3.05   

   Total Number of Purchases Total Revenue for Pymoli  
0                        780                $2,379.77

In [6]:
#Gender Demographics

#Percentage and Count of Male Players
#Percentage and Count of Female Players
#Percentage and Count of Other / Non-Disclosed

In [7]:
#verifiying total count will be accurate 
#does not exclude duplicates...total players from above is only 573
datagen_count = pymoli_data['Gender'].value_counts()
datagen_count


Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [8]:
# testing and verifying unquie total. 
# 484+81+11 = 576
gender_nonduplicate = pymoli_data.drop_duplicates(['SN'])
true_totalcount = gender_nonduplicate['Gender'].count()

gender_nonduplicate['Gender'].value_counts()

Male                     484
Female                    81
Other / Non-Disclosed     11
Name: Gender, dtype: int64

In [9]:
#total count...in the gender column, keep rows with 'Male', 'Female', 'Others' and count them
male_count = gender_nonduplicate['Gender'].loc[gender_nonduplicate['Gender']=='Male'].count()
female_count = gender_nonduplicate['Gender'].loc[gender_nonduplicate['Gender']=='Female'].count()
others_count = gender_nonduplicate['Gender'].loc[gender_nonduplicate['Gender']=='Other / Non-Disclosed'].count()

#gender percentage
male_percent = (male_count / true_totalcount) * 100
female_percent = (female_count / true_totalcount) * 100
other_percent = (others_count / true_totalcount) * 100

#summary table for gender
gender_summarydf = {'Total Count': [male_count, female_count, others_count],\
                       'Players Percentage': [male_percent, female_percent, other_percent]}
gender_summary = pd.DataFrame(data=gender_summarydf, index = ['Male', 'Female', 'Other / Non-Disclosed'],\
                              columns = ['Total Count', 'Players Percentage'])

gender_summary['Players Percentage'] = gender_summary['Players Percentage'].map("{0:,.2f}%".format)

gender_summary

Total Count Players Percentage
Male                           484             84.03%
Female                          81             14.06%
Other / Non-Disclosed           11              1.91%

In [10]:
#Purchasing Analysis (Gender)

#The below each broken by gender

#1 Purchase Count
#2 Average Purchase Price
#3 Total Purchase Value
#4 Average Purchase Total per Person by Gender

In [38]:
item_df = pymoli_data.set_index('Item ID')

#1 Purchase Count...using gender column, keep all rows that has 'Male', 'Female', 'Others' and count the row to get how many 
#items were purchased by the gender group
male_items = item_df['Gender'].loc[item_df['Gender'] == 'Male'].count()
female_items = item_df['Gender'].loc[item_df['Gender'] == 'Female'].count()
other_items = item_df['Gender'].loc[item_df['Gender'] == 'Other / Non-Disclosed'].count()

#2 Average Purchase Price...with the column 'Price', keep rows that has 'Male', 'Female', 'Others', and sum the numbers 
#in the Price column and then divide that by the count above
avg_price_4male = (item_df['Price'].loc[item_df['Gender'] == 'Male'].sum())/male_items
avg_price_4female = (item_df['Price'].loc[item_df['Gender'] == 'Female'].sum())/female_items
avg_price_4other = (item_df['Price'].loc[item_df['Gender'] == 'Other / Non-Disclosed'].sum())/other_items

#3 Total Purchase Value...rows with 'Male', 'Female', 'Others', sum the values in the 'Price' column
purch_4male = item_df['Price'].loc[item_df['Gender'] == 'Male'].sum()
purch_4female = item_df['Price'].loc[item_df['Gender'] == 'Female'].sum()
purch_4other = item_df['Price'].loc[item_df['Gender'] == 'Other / Non-Disclosed'].sum()

#4 Average Purchase Total per Person by Gender
total_maleavg = purch_4male/male_count
total_femavg = purch_4female/female_count
total_otheravg = purch_4other/others_count

#summary table
purchasing_summary = {'Purchase Count': [male_items, female_items, other_items],\
                    'Average Purchase Price': [avg_price_4male, avg_price_4female, avg_price_4other],\
                    'Total Purchase Value': [purch_4male, purch_4female, purch_4other],\
                    'Avg Total Purchase per Person': [total_maleavg, total_femavg, total_otheravg]}

purchase_summary_df = pd.DataFrame(data=purchasing_summary, index = ['Male', 'Female', 'Other / Non-Disclosed'],\
                              columns = ['Purchase Count', 'Average Purchase Price', 'Total Purchase Value', 'Avg Total Purchase per Person'])
purchase_summary_df

# need to add some formatting to make it look table look nice
purchase_summary_df['Average Purchase Price'] = purchase_summary_df['Average Purchase Price'].map("${0:,.2f}".format)
purchase_summary_df['Avg Total Purchase per Person'] = purchase_summary_df['Avg Total Purchase per Person'].map("${0:,.2f}".format)

purchase_summary_df

Purchase Count Average Purchase Price  \
Male                              652                  $3.02   
Female                            113                  $3.20   
Other / Non-Disclosed              15                  $3.35   

                       Total Purchase Value Avg Total Purchase per Person  
Male                                1967.64                         $4.07  
Female                               361.94                         $4.47  
Other / Non-Disclosed                 50.19                         $4.56

In [12]:
#Age Demographics

#1 The below each broken into bins of 4 years (i.e. <10, 10-14, 15-19, etc.)

#2 
#3 
#4 
#5 Average Purchase Total per Person by Age Group

In [14]:
#1 The below each broken into bins of 4 years (i.e. <10, 10-14, 15-19, etc.)
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]
age_group = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']

#2 Purchase Count
pymoli_data['Category of players by age'] = pd.cut(gender_nonduplicate['Age'], bins, labels=age_group )

category_count = pymoli_data['Category of players by age'].value_counts()

#3 Average Purchase Price
#4 Total Purchase Value
#5 Average Purchase Total per Person by Age Group
percent_4players = category_count / total_players *100

age_demo_summary = {'Total Count': category_count, 'Percentage of Players': percent_4players}
age_demo_summarydf = pd.DataFrame(data=age_demo_summary, index=age_group, columns=['Total Count', 'Percentage of Players'])
age_demo_summarydf

age_demo_summarydf['Percentage of Players'] = age_demo_summarydf['Percentage of Players'].map("{0:,.2f}%".format)
age_demo_summarydf

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

In [15]:
#Age Demographics

#1 The below each broken into bins of 4 years (i.e. <10, 10-14, 15-19, etc.)

#2 Purchase Count
#3 Average Purchase Price
#4 Total Purchase Value
#5 Average Purchase Total per Person by Age Group

In [20]:
#1 The below each broken into bins of 4 years (i.e. <10, 10-14, 15-19, etc.)
bins_2 = [0, 9, 14, 19, 24, 29, 34, 39, 100]
age_group_2 = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']

#reset index to replace it with Category of players by age
tableindex_reset = pymoli_data.reset_index(drop=True)

tableindex_reset['Category of players by age'] = pd.cut(tableindex_reset['Age'], bins_2, labels=age_group_2 )

#2 Purchase Count
#3 Average Purchase Price
#4 Total Purchase Value
#5 Average Purchase Total per Person by Age Group
category_bybins = tableindex_reset.groupby('Category of players by age')
category_counts2 = category_bybins['Purchase ID'].count()
category_average2 = category_bybins['Price'].mean()
category_total2 = category_bybins['Price'].sum()

#like total players, need to remove duplicates otherwise it will mess with the average
dups_drop_4category = tableindex_reset.drop_duplicates(['SN'])
correctcount=dups_drop_4category['SN'].count()

x= category_bybins['SN'].count()
y = category_bybins['Price'].sum() 
category_average = y / x

#summary table building
age_demo_df = pd.DataFrame(category_counts2)
age_demo_df = age_demo_df.rename(columns={'Purchase ID': 'Purchase Count'})
age_demo_df['Average Purchase Price'] = category_average2
age_demo_df['Total Purchase Value'] = category_total2
age_demo_df['Avg Total Purchase per Person'] = category_average

age_demo_df

#needs formatting like previous summary table 
age_demo_df['Average Purchase Price'] = age_demo_df['Average Purchase Price'].map("${0:,.2f}".format)
age_demo_df['Total Purchase Value'] = age_demo_df['Total Purchase Value'].map("${0:,.2f}".format)
age_demo_df['Avg Total Purchase per Person'] = age_demo_df['Avg Total Purchase per Person'].map("${0:,.2f}".format)


age_demo_df

Purchase Count Average Purchase Price  \
Category of players by age                                          
<10                                     23                  $3.35   
10-14                                   28                  $2.96   
15-19                                  136                  $3.04   
20-24                                  365                  $3.05   
25-29                                  101                  $2.90   
30-34                                   73                  $2.93   
35-39                                   41                  $3.60   
40+                                     13                  $2.94   

                           Total Purchase Value Avg Total Purchase per Person  
Category of players by age                                                     
<10                                      $77.13                         $3.35  
10-14                                    $82.78                         $2.96  
15-19                                   $412.89                         $3.04  
20-24                                 $1,114.06                         $3.05  
25-29                                   $293.00                         $2.90  
30-34                                   $214.00                         $2.93  
35-39                                   $147.67                         $3.60  
40+                                      $38.24                         $2.94

In [ ]:
#Top Spenders

#Identify the the top 5 spenders in the game by total purchase value, then list (in a table):

#SN
#Purchase Count
#Average Purchase Price
#Total Purchase Value

In [25]:
# Group by SN to calculate total purchase count. i will need to go off the previous table where purchase was calculated
topspender_purch = tableindex_reset.groupby('SN')
topspender = topspender_purch['Item ID'].count()
topspender_avg = topspender_purch['Price'].mean()
topspender_avg_overall = topspender_purch['Price'].sum()

topspender_purch = pd.DataFrame(topspender)

# Renaming Purchase ID columns to Purchase count
topspender_purch = topspender_purch.rename(columns={'Item ID': 'Purchase Count'})

#building summary table
topspender_purch['Average Purchase Price'] = topspender_avg
topspender_purch['Total Purchase Value'] = topspender_avg_overall

#filter through table to have highest spender on top
topspender_purch2= topspender_purch.sort_values(['Total Purchase Value'], ascending=False)
topspender_purch2.head(1)


topspender_purch2['Average Purchase Price'] = topspender_purch2['Average Purchase Price'].map("${0:,.2f}".format)
topspender_purch2['Total Purchase Value'] = topspender_purch2['Total Purchase Value'].map("${0:,.2f}".format)
topspender_purch2.head(1)


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                   
Lisosia93               5                  $3.79               $18.96

In [26]:
#Most Popular Items

#Identify the 5 most popular items by purchase count, 
#then list (in a table):

#Item ID
#Item Name
#Purchase Count
#Item Price
#Total Purchase Value

In [34]:
#reset original
popular_item = tableindex_reset[['Item ID', 'Item Name', 'Price']]

popular_item_df = popular_item.groupby(['Item ID', 'Item Name'])

pop_purc_count = popular_item_df['Item ID'].count()
total_count = popular_item_df['Price'].sum()
pop_item_price = total_count/pop_purc_count

popular_df = pd.DataFrame(pop_purc_count)

popular_df = popular_df.rename(columns={'Item ID': 'Purchase Count'})
popular_df['Item Price']=pop_item_price
popular_df['Total Purchase Value'] = total_count

sorted_pop_item = popular_df.sort_values(['Purchase Count'], ascending=False)

sorted_pop_item

sorted_pop_item['Item Price'] = sorted_pop_item['Item Price'].map("${0:,.2f}".format)
sorted_pop_item['Total Purchase Value'] = sorted_pop_item['Total Purchase Value'].map("${0:,.2f}".format)

sorted_pop_item.head(1)


Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76

In [ ]:
#Most Profitable Items

#Identify the 5 most profitable items by total purchase value, then list (in a table):

#Item ID
#Item Name
#Purchase Count
#Item Price
#Total Purchase Value

In [36]:
profitable_items = popular_df.sort_values(['Total Purchase Value'], ascending=False)

profitable_items

profitable_items['Item Price'] = profitable_items['Item Price'].map("${0:,.2f}".format)
profitable_items['Total Purchase Value'] = profitable_items['Total Purchase Value'].map("${0:,.2f}".format)

profitable_items.head(1)

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76

Trend Conclusion: The game Heroes of Pymoli is prominently popular among men, for over 84% of the playerbase, 576 players, identified as male. Meanwhile, over 14% of the playerbase identified as female and almost 2% identified as others/non-disclosed.
Out of all the players, Heroes of Pymoli is most popular with those between 20-24 years old. Although, players between 20 - 24 are the most popluar age group of Heroes of Pymoli, those between the age of 35-39 spends the most on average on the game at $3.60     